scipy/stats/_stats_py.py

```diff
@@ -3191,6 +3210,51 @@ def zmap(scores, compare, axis=0, ddof=0, nan_policy='propagate'):
     array([-1.06066017,  0.        ,  0.35355339,  0.70710678])
 
     """
+    # The docstring explicitly states that it preserves subclasses.
+    # Let's table deprecating that and just get the array API version
+    # working.
+    implementation = _zmap_array_api if SCIPY_ARRAY_API else _zmap_not_array_api
+
+    with warnings.catch_warnings():
+        if scores is compare:  # zscore should not emit SmallSampleWarning
+            warnings.simplefilter('ignore', SmallSampleWarning)
+
+        return implementation(scores, compare, axis, ddof, nan_policy)
+
+
+def _convert_common_float(*arrays, xp=None):
+    xp = array_namespace(*arrays) if xp is None else xp
+    arrays = [xp.asarray(array) for array in arrays]
```

Comment:
I think this might need to be
```suggestion
    arrays = [_asarray(array, subok=True) for array in arrays]
```
otherwise line 3158 will break subclasses for `gzscore`.

Also with this change `_zmap_array_api` will preserve subclasses so could we just keep the new shorter implementation or is the nan behaviour backwards incompatible?

Reply:
Mmm maybe. I'll investigate. But in the meantime I've followed the approach of `zmap` and just branched here depending on `SCIPY_ARRAY_API` to avoid any possible unintended interactions.

Reply:
@j-bowhay the second to last commit removes the old implementation. Besides using `_asarrray`, I needed to replace a use of `xp.where`, which doesn't preserve array subclasses. 

Note that `_xp_mean` and `_xp_var` "just worked" as far as the existing unit tests are concerned, but note that they weren't written with masked arrays in mind, and the unit tests are very weak. I don't have a preference if you're willing to test/review more carefully to check that removing the old implementation is safe (including for other subclasses, like `matrix`), but personally, I'd just as soon keep both implementations and remove the old one when we no longer have to support subclasses.

Reply:
Ok thanks I will look more closely over the next couple of days